README
libicd-network-usb
==================

libicd-network-usb provides a network module for ICd2 that controls USB network
devices. The network module will look up the interface index for a network
interface named 'usb0'. The USB network interface is enabled by attaching an
USB cable and selecting 'PC Suite mode'. The module will detect the presence
of an USB network as long as the g_nokia kernel module is loaded. It will also
work if the device is in host mode, has a USB ethernet dongle attached and
proper kernel modules loaded for that ethernet dongle.

When a network scan is performed by icd2, the USB network module will list all
networks configured in GConf with of type 'USB', provided that the 'usb0'
network interface is found. When a USB network is selected from the 'Select
connection' dialog, the same check for 'usb0' will be performed and success
reported to icd2. Thus the next network module, in this case
libicd-network-ipv4 will be run, which then will attempt to configure an
IPv4 address using DHCP by default. It is up to the other end - PC or real
network - to provide the required DHCP server.

For the USB nework to work, it needs to be added to GConf with:
gconftool-2 -s -t string /system/osso/connectivity/IAP/<configname>/type USB
gconftool-2 -s -t string /system/osso/connectivity/IAP/<configname>/name 'Usb network'

Optionally it is possible to set the autoconnect flag with:
gconftool-2 -s -t boolean /system/osso/connectivity/IAP/<configname>/autoconnect true

libicd-network-ipv4 will work normally, i.e. use its IP configuration options
from the same /system/osso/connectivity/IAP/<configname>/ GConf path. The
connectivity UI seems to try to configure a WLAN network connection if the
settings are edited in the control panel in Fremantle. IP address selection
parameters are located behind the 'Advanced' button on the last page. After any
editing via the UI please check that the network type is properly (re)set to
'USB' by using the command above. Any other WLAN configuration items left in
GConf are ignored by the USB network module.


debian/changelog
libicd-network-usb (0.3.1) unstable; urgency=medium

  * Port to debhelper 9 [Ivaylo Dimitrov]

 -- Ivan J. <parazyd@dyne.org>  Sun, 11 Mar 14:05:40 +0100

libicd-network-usb (0.3) unstable; urgency=low

  * Fix one more leftover autoconf dependency

 -- Patrik Flykt <patrik.flykt@nokia.com>  Fri, 19 Feb 2010 09:36:41 +0200

libicd-network-usb (0.2) unstable; urgency=low

  * Fixed leftover dependency preventing build discovered by
    thresh[at]altlinux.ru.

 -- Patrik Flykt <patrik.flykt@nokia.com>  Wed, 17 Feb 2010 13:09:32 +0200

libicd-network-usb (0.1) unstable; urgency=low

  * USB network module based on dummy network module

 -- Patrik Flykt <patrik.flykt@nokia.com>  Thu, 11 Dec 2010 11:26:00 +0200



src/libicd_network_usb.h
/*
 * This file is part of libicd-network-usb
 *
 * Copyright (C) 2010 Nokia Corporation.
 *
 * Contact: Patrik.Flykt@nokia.com <patrik.flykt@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */

#ifndef LIBICD_NETWORK_USB
#define LIBICE_NETWORK_USB

/** Network type */
#define USB_NW_TYPE      "USB"
/** The (one and only) USB network interface we support */
#define USB_NW_INTERFACE "usb0"

#endif


src/libicd_network_usb.c
/*
 * This file is part of libicd-network-usb
 *
 * Copyright (C) 2010 Nokia Corporation.
 *
 * Contact: Patrik.Flykt@nokia.com <patrik.flykt@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <stdio.h>
#include <errno.h>
#include <sys/ioctl.h>
#include <net/if.h>

#include <string.h>
#include <glib.h>
#include <gconf/gconf-client.h>

#include <osso-ic-gconf.h>
#include <network_api.h>

#include "libicd_network_usb.h"


gboolean icd_nw_init (struct icd_nw_api *network_api,
		      icd_nw_watch_pid_fn watch_cb,
		      gpointer watch_cb_token,
		      icd_nw_close_fn close_cb);


/** Check whether the given interface exists
 * @param ifname the interface name
 * @return TRUE if #USB_NW_INTERFACE is up, FALSE otherwise
 */
static gboolean usb_if_exists (const gchar *ifname)
{
  int fd;
  struct ifreq iface;
  gboolean found = FALSE;

  if ((fd = socket (AF_UNIX, SOCK_DGRAM, 0)) < 0)
    return found;

  strcpy (iface.ifr_name, ifname);
  if (ioctl (fd, SIOCGIFINDEX, &iface) == 0)
    found = TRUE;

  close (fd);
  return found;
}


/** Function to bring up the link layer
 * @param network_type network type
 * @param network_attrs attributes, such as type of network_id, security, etc.
 * @param network_id IAP name or local id, e.g. SSID
 * @param link_up_cb the callback function to call when the link is up
 * @param link_up_cb_token token to pass to the callback function
 * @param private a reference to the icd_nw_api private memeber
 */
static void usb_link_up (const gchar *network_type,
			   const guint network_attrs,
			   const gchar *network_id,
			   icd_nw_link_up_cb_fn link_up_cb,
			   const gpointer link_up_cb_token,
			   gpointer *private)
{
  if (usb_if_exists (USB_NW_INTERFACE)) {
    link_up_cb (ICD_NW_SUCCESS,
		NULL,
		USB_NW_INTERFACE,
		link_up_cb_token, NULL);
  } else {
    link_up_cb (ICD_NW_ERROR,
		NULL,
		NULL,
		link_up_cb_token, NULL);
  }
}

/** Function for listing the available networks provided by the module
 * @param network_type network type or NULL
 * @param search_scope search scope; ignored
 * @param search_cb the search callback
 * @param search_cb_token token from the ICd to pass to the callback
 * @private a reference to the icd_nw_api private member
 */
static void usb_start_search (const gchar *network_type,
				guint search_scope,
				icd_nw_search_cb_fn search_cb,
				const gpointer search_cb_token,
				gpointer *private)
{
  GConfClient *gconf_client;
  GSList *list;
  gchar *gconf_path, *gconf_name, *key, *settings_type, *settings_name;
  gboolean settings_autoconnect;


  if (usb_if_exists(USB_NW_INTERFACE)) {

    gconf_client = gconf_client_get_default();
    list = gconf_client_all_dirs (gconf_client,
				  ICD_GCONF_PATH,
				  NULL);

    while (list) {
      gconf_path = list->data;

      /* IAP id */
      gconf_name = g_strrstr (gconf_path, "/");
      if (gconf_name) {
	gconf_name += 1;
	gconf_name = gconf_unescape_key (gconf_name, -1);
      }

      /* type */
      key = g_strconcat (gconf_path, "/type", NULL);
      settings_type = gconf_client_get_string (gconf_client, key, NULL);
      g_free (key);

      if (gconf_name &&
	  settings_type &&
	  strcmp (settings_type, USB_NW_TYPE) == 0) {

	/* autoconnect attribute */
	key = g_strconcat (gconf_path, "/autoconnect", NULL);
	settings_autoconnect = gconf_client_get_bool (gconf_client, key, NULL);
	g_free (key);

	/* IAP user readable name */
	key = g_strconcat (gconf_path, "/name", NULL);
	settings_name = gconf_client_get_string (gconf_client, key, NULL);
	g_free (key);
	if (settings_name == NULL)
	  settings_name = g_strdup (gconf_name);
					       
	search_cb (ICD_NW_SEARCH_CONTINUE,
		   settings_name,
		   USB_NW_TYPE,
		   ICD_NW_ATTR_IAPNAME |
		   (settings_autoconnect? ICD_NW_ATTR_AUTOCONNECT: 0),
		   gconf_name,
		   ICD_NW_LEVEL_10,
		   /** TODO: add MAC address here */
		   "",
		   0,
		   search_cb_token);
	g_free (settings_name);
      }
      g_free (gconf_name);
      g_free (settings_type);
      g_free (list->data);
      list = g_slist_delete_link (list, list);
    }
    g_object_unref (gconf_client);
  }

  search_cb (ICD_NW_SEARCH_COMPLETE,
	     NULL,
	     NULL,
	     0,
	     NULL,
	     0,
	     NULL,
	     0,
	     search_cb_token);
}

/** USB network module initialization function.
 * @param network_api icd_nw_api structure filled in by the module
 * @param watch_cb function to inform ICd that a child process is to be
 *        monitored for exit status
 * @param watch_cb_token token to pass to the watch pid function
 * @param close_cb function to inform ICd that the network connection is to be
 *        closed
 * @return TRUE on succes; FALSE on failure whereby the module is unloaded
 */
gboolean icd_nw_init (struct icd_nw_api *network_api,
		      icd_nw_watch_pid_fn watch_cb,
		      gpointer watch_cb_token,
		      icd_nw_close_fn close_cb)
{
  network_api->version = ICD_NW_MODULE_VERSION;
  network_api->link_up = usb_link_up;
  network_api->search_lifetime = 31;
  network_api->search_interval = 30;
  network_api->start_search = usb_start_search;

  return TRUE;
}
